docs: blog post for sbom epic (#3216)
open-component-model/open-component-model: website/content/blog/2026-07-28-shipping-sboms-with-your-components.md

---
title: "Shipping SBOMs with Your Components"
description: "A proof of concept for transporting Software Bills of Materials alongside OCM components — linking, discovering, and orchestrating SBOMs so a whole component version could be scanned in one command."
date: 2026-07-28T10:00:00+02:00
contributors: []
tags: ["supply-chain", "sbom", "security", "cli"]
draft: false
---

## The Problem

> OCM already tells you *which artifacts* a component delivers — its Software Bill of Delivery. But that does not answer the question a security team actually has: *"which dependencies are bundled inside those artifacts, and am I affected by this vulnerability?"* Knowing a component ships `podinfo:6.9.1` and a CLI binary says nothing about the packages and libraries buried inside them.

Reviewer: ```suggestion
OCM already tells you *which artifacts* a component delivers — its Software Bill of Delivery. But that does not answer the question a security team actually has: *"which dependencies are bundled inside those artifacts, and am I affected by this vulnerability?"* Knowing a component `podinfo:6.9.1` contains a CLI binary as a resource, does not tell anything about the packages and libraries buried inside that resource.
```

Author: ```suggestion
OCM already tells you *which artifacts* a component delivers — its Software Bill of Delivery. But that does not answer the question a security team actually has: *"which dependencies are bundled inside those artifacts, and am I affected by this vulnerability?"* Knowing a component ships an OCI image and a CLI binary as a resource, does not tell anything about the packages and libraries buried inside that resource.
```

> OCM already tells you *which artifacts* a component delivers — its Software Bill of Delivery. But that does not answer the question a security team actually has: *"which dependencies are bundled inside those artifacts, and am I affected by this vulnerability?"* Knowing a component ships an OCI image and a CLI binary as a resource, does not tell anything about the packages and libraries buried inside that resource.

Author: fixed

Those answers live in Software Bills of Materials (SBOMs) — but today they are scattered: natively attached to an individual OCI artifact (so you have to know to look there), added to the component but with no relation to the resource they describe, or transported entirely outside the component, where they drift out of sync with what actually shipped and lose their trustworthiness. There is rarely one trustworthy answer for a whole component.

> The idea explored here: make SBOMs travel *with* the component, related to the artifacts they describe, so the dependency-and-vulnerability question can be answered straight from the delivered component — online or air-gapped, and covered by the same trust as the rest of the component.

Reviewer: ```suggestion
The idea explored here: make SBOMs travel *with* the component, related to the artifacts they describe, so the dependency-and-vulnerability question can be answered straight from the delivered component — online or air-gapped, and covered by the same trust as the rest of the signed component.
```

---

## What We Prototyped

To explore the idea, we built a proof of concept across three areas. None of this is a shipped feature yet — it is a spike to see what the workflow could feel like, and one we are now actively driving forward.

**Linking an SBOM to what it describes.** A resource of `type: sbom` carries an `ocm.software/artefactReference` label whose `identitySelector` points at the resource the SBOM is about. Because it is an ordinary, signed resource, the link travels and is verified with the component — no side channel.

```yaml
- name: cli-sbom
  type: sbom
  labels:
    - name: ocm.software/artefactReference
      version: v1
      value:
        identitySelector:
          name: cli        # the resource this SBOM describes
      signing: true
  input:
    type: File/v1
    path: ./cli.sbom.json
    mediaType: application/vnd.cyclonedx+json
```

> **Discovering SBOMs that already exist.** Many images already ship an SBOM (buildx attestations, OCI Referrers). A prototype `SBoM/v1` input discovers that SBOM at build time and links it to the component — *discover, never generate* — so it works with artifacts as they are, in their original format.

Reviewer: ```suggestion
**Discovering SBOMs that already exist.** Many images already ship an SBOM (Docker buildx attestations, OCI Referrers). A prototype `SBoM/v1` input discovers that SBOM at build time and links it to the component — *discover, never generate* — so it works with artifacts as they are, in their original format.
```

```yaml
- name: podinfo-sbom
  type: sbom
  input:
    type: SBoM/v1
    resource:
      name: podinfo       # discover the SBOM attached to this image
```

**Consuming the result.** A single SBOM is just a normal resource download. The interesting part is the *orchestrating* SBOM: one command assembles every resource's SBOM into a single CycloneDX document (with `--recursive`, across referenced components too), which you can hand straight to a scanner:

```shell
ocm get sbom ghcr.io/org/product:v1 --recursive -o json > product.cdx.json
trivy sbom product.cdx.json
```

```text
┌────────┬──────────┬─────────────────┐
│ Target │   Type   │ Vulnerabilities │
├────────┼──────────┼─────────────────┤
│        │ gobinary │       58        │
└────────┴──────────┴─────────────────┘
```

> One command produces a bill of materials for the whole component; the next answers "am I affected?" — from the delivered component, online or air-gapped, with no per-artifact hunting. We also explored teaching [OpenDeliveryGear (ODG)](https://open-component-model.github.io/open-delivery-gear/) to consume the SBOMs (with VEX) directly, so that vendor-provided triage data is applied automatically.

Reviewer: 1. "produces a bill of materials" or SBOM?
2. "to consume the SBOMs (with VEX) directly" - is it VEX that enables the consumption (meaning with the help of VEX) or is it an SBOM containing / referencing a VEX document?

Author: improved the wording
